### PR TITLE
Fix 32-bit overflows in LLVM composite constants

### PR DESCRIPTION
**Pull request by erer1243**
Inspired by #121868. Fixes unsoundness created when constructing constant arrays, strings, and structs with 2^32 or more elements on x86_64. This introduces copies of a few LLVM functions that have their signatures updated to use size_t in place of unsigned int. Alternatively we could just add overflow checks and just disallow huge composite constants. That introduces less code, but maybe a huge static block of memory is useful in embedded/no-os situations?

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/common.rs [hunk 1]
```diff
@@ -108,7 +108,11 @@ impl<'ll> CodegenCx<'ll, '_> {
 
     pub fn const_get_elt(&self, v: &'ll Value, idx: u64) -> &'ll Value {
         unsafe {
-            assert_eq!(idx as c_uint as u64, idx);
+            assert!(
+                idx <= c_uint::MAX as u64,
+                "LLVMGetAggregateElement doesn't support indices greater than {}",
+                c_uint::MAX
+            );
             let r = llvm::LLVMGetAggregateElement(v, idx as c_uint).unwrap();
```
Comment (kadiwa4) [inline compiler/rustc_codegen_llvm/src/common.rs:116]: Could you use `c_uint::try_from` here? Seems cleaner than an `as` cast

Comment (erer1243) [reply compiler/rustc_codegen_llvm/src/common.rs:116]: sure

**Comment by nnethercote** [comment]
Not my area of expertise, let's try r? @nikic

**Comment by nikic** [comment]
Would you be interested in upstreaming the new APIs to LLVM? If you do that, then we can directly start using them by providing shims like this:
```
// Or whatever version it was added
#if LLVM_VERSION_LE(17, 0)
LLVMValueRef LLVMConstArray2(LLVMTypeRef ElementTy, LLVMValueRef *ConstantVals,
                             uint64_t Length) {
  ArrayRef<Constant *> V(unwrap<Constant>(ConstantVals, Length), Length);
  return wrap(ConstantArray::get(ArrayType::get(unwrap(ElementTy), Length), V));
}
#endif
```

That way we'll automatically switch to using the upstream APIs as soon as we can.

**Comment by erer1243** [comment]
@nikic Yeah, I was planning on doing that. I'm not sure how long it would take to get through, so I'm not sure which version we would put in the version macro.

**Comment by erer1243** [comment]
Having 64bit APIs for structs and vectors did not make sense, because LLVM stores 32bit lengths for them anyway. Now we just assert for overflows before calling those functions. I made a pr to LLVM for 64bit ConstantString constructors and used your idea @nikic to create shims to delete later. If the LLVM patch gets through quickly we can update the LLVM_VERSION_LT check to be 19.

**Comment by erer1243** [comment]
Barring the LLVM patch being rescinded, this is now finished.

File: compiler/rustc_codegen_llvm/src/llvm/ffi.rs [hunk 2]
```diff
@@ -932,10 +932,16 @@ extern "C" {
     pub fn LLVMConstReal(RealTy: &Type, N: f64) -> &Value;
 
     // Operations on composite constants
-    pub fn LLVMConstStringInContext(
+    pub fn LLVMConstArray2<'a>(
+        ElementTy: &'a Type,
+        ConstantVals: *const &'a Value,
+        Length: size_t,
```
Comment (nikic) [inline compiler/rustc_codegen_llvm/src/llvm/ffi.rs:938]: This should be u64.

Comment (erer1243) [reply compiler/rustc_codegen_llvm/src/llvm/ffi.rs:938]: oops thanks

File: compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp [hunk 3]
```diff
@@ -2079,3 +2072,36 @@ extern "C" bool LLVMRustLLVMHasZlibCompressionForDebugSymbols() {
...
   return llvm::compression::zstd::isAvailable();
 }
+
+// Operations on composite constants.
+// These are clones of LLVM api functions that will become available in future releases.
+// They can be removed once Rust's minimum supported LLVM version supports them.
+// See https://github.com/rust-lang/rust/issues/121868
+// See https://llvm.org/doxygen/group__LLVMCCoreValueConstantComposite.html
+
+// FIXME: Remove when RUST's minimum supported LLVM version reaches 19.
```
Comment (nikic) [inline compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:2082]:
```suggestion
// FIXME: Remove when Rust's minimum supported LLVM version reaches 19.
```

**Review by nikic — COMMENTED** [review]
Looks good to me now. Could you please squash the commits?

**Pull request merged**